### PR TITLE
added buidler deploy plug in and updated instantiate contract task

### PR DESCRIPTION
**Pull request by hilmarx**
**Relevant changes:**

- Added builder deploy plugin. Still need to update all tests with it (outstanding), not prio at the moment
- Added 2 deploy scripts, one setting up GelatoCore, one GelatoUserProxies
- Due to buider deploy script, I had to rename the old `deploy` task to `gc-deploy`
- Updated addresses of executor for mainnet and rinkeby
- added some helper tasks

### CODE REVIEW

File: deploy/setupGelatoUserProxies.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    log(
+      `// ==== GelatoActionPipeline deployed => ${GelatoActionPipeline.address} ====`
+    );
+    log(
+      `// ==== GelatoActionPipeline deployed => ${ProviderModuleGelatoUserProxy.address} ====`
+    );
+  }
+};
+
+module.exports.tags = ["ProviderModules"];
```
Comment (gitpusha) [inline deploy/setupGelatoUserProxies.js:53]: Why "ProviderModules" and not "GelatoUserProxyFactory", "GelatoActionPipeline", "ProviderModuleGelatoUserProxy" ?

**Comment by gitpusha** [comment]
I guess my general problem with this PR is that I don't really understand what the file changes are good for yet. To me it looks a bit like unnecessary commits to a `master` that was supposed to stabilize.

But probably you have some usage in mind that I cannot see just from the file changes. 

Is this PR basically there so that in the future we can deploy new contract with `buidler-deploy` ? I guess that would make sense then. 

> - Added builder deploy plugin. Still need to update all tests with it (outstanding), not prio at the moment

I don't see the value-add. We should keep `master` as stable as possible. And this PR does not really add any major required features to `master`. It only adds a dependency to package.json which is barely even used anywhere. I am reluctant to merge half-measures into `master`.

> - Added 2 deploy scripts, one setting up GelatoCore, one GelatoUserProxies

Again, I don't see the value add. This is basically redundant because we already had scripts to do this and this PR doesnt add any important usage or improvements from adding `buidler-deploy` as a dependency.

> Due to buider deploy script, I had to rename the old deploy task to gc-deploy
I would call it `custom-deploy`. `gc` is GelatoCore. But again. If we don't make extensive use of the `buidler-deploy` dependency, why add it as a half-measure?

> Updated addresses of executor for mainnet and rinkeby
See my commit suggestion for up to date address of executorModule on mainnet

File: buidler/tasks/gelato/core/task_receipts/iTask.fetchTaskReceipt.js [hunk 2]
```diff
@@ -49,7 +49,7 @@ export default task(
           contractname: "GelatoCore",
           contractaddress: taskArgs.contractaddress,
           eventname: "LogTaskSubmitted",
-          fromblock: taskArgs.fromblock,
+          fromblock: 0,
```
[the remark this answers is not in the document]
Comment (hilmarx) [reply buidler/tasks/gelato/core/task_receipts/iTask.fetchTaskReceipt.js:52]: I had weird bug that, when using `taskArgs.fromblock`, did not show all available task receipts. Inputting 0 solved that.

File: deploy/setupGelatoUserProxies.js (hunk 1 above)
Comment (hilmarx) [inline deploy/setupGelatoUserProxies.js:53]:
```suggestion
module.exports.tags = ["GelatoUserProxyFactory", "GelatoActionPipeline", "ProviderModuleGelatoUserProxy"];
```

File: buidler/tasks/erc20/task.dai-rinkeby-faucet.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { task } from "@nomiclabs/buidler/config";
+import { defaultNetwork } from "../../../buidler.config";
+
+export default task(
+  "get-dai",
+  `Return (or --log) <erc20address> allowance by <owner> to <spender> on [--network] (default: ${defaultNetwork})`
```
Comment (hilmarx) [inline buidler/tasks/erc20/task.dai-rinkeby-faucet.js:6]:
```suggestion
  `DAI faucet for rinkeby - mints 100 DAI to User Account`
```

File: buidler/tasks/erc20/task.dai-rinkeby-faucet.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    const dai = await ethers.getContractAt(
+      daiAbi,
+      "0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa"
+    );
+    console.log(dai);
+    console.log("to here");
+    const tx = await dai.allocateTo(
+      userAddress,
+      1 // ethers.utils.parseUnits("100", 18)
```
Comment (hilmarx) [inline buidler/tasks/erc20/task.dai-rinkeby-faucet.js:24]:
```suggestion
      ethers.utils.parseUnits("100", 18)
```

File: buidler/tasks/erc20/task.dai-rinkeby-faucet.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    const daiAbi = [
+      "function allocateTo(address _userAddress, uint256 _amount)",
+    ];
+
+    const dai = await ethers.getContractAt(
+      daiAbi,
+      "0x5592EC0cfb4dbc12D3aB100b257153436a1f0FEa"
+    );
+    console.log(dai);
+    console.log("to here");
```
Comment (hilmarx) [inline buidler/tasks/erc20/task.dai-rinkeby-faucet.js:21]:
```suggestion
```

File: buidler/tasks/erc20/task.dai-rinkeby-faucet.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    const tx = await dai.allocateTo(
+      userAddress,
+      1 // ethers.utils.parseUnits("100", 18)
+    );
+    console.log("bug");
```
Comment (hilmarx) [inline buidler/tasks/erc20/task.dai-rinkeby-faucet.js:26]:
```suggestion
```

File: buidler/tasks/gelato/core/gelato_providers/provider_module/example_provider_modules/task.deployGnosisSafeModule.js [hunk 7]
```diff
@@ -3,7 +3,7 @@ import { defaultNetwork } from "../../../../../../../buidler.config";
 import { utils, constants } from "ethers";
 
 export default task(
-  "gc-deploy-gnosis-safe-module",
+  "deploy-gnosis-safe-module",
```
Comment (hilmarx) [inline buidler/tasks/gelato/core/gelato_providers/provider_module/example_provider_modules/task.deployGnosisSafeModule.js:6]:
```suggestion
  "gc-deploy-gnosis-safe-module",
```

File: buidler/tasks/internal/contracts/iTask.instantiateContract.js [hunk 8]
```diff
@@ -30,10 +29,12 @@ export default internalTask(
 
       let instance;
       if (read) {
```
Comment (hilmarx) [inline buidler/tasks/internal/contracts/iTask.instantiateContract.js:31]:
```suggestion
```

File: buidler/tasks/internal/contracts/iTask.instantiateContract.js [hunk 9]
```diff
@@ -30,10 +29,12 @@ export default internalTask(
...
-        instance = new Contract(contractaddress, abi, ethers.provider);
+        // instance = new Contract(contractaddress, abi, ethers.provider);
+        instance = await ethers.getContractAt(abi, contractaddress);
       } else if (write) {
```
Comment (hilmarx) [inline buidler/tasks/internal/contracts/iTask.instantiateContract.js:34]:
```suggestion
```

File: buidler/tasks/internal/contracts/iTask.instantiateContract.js [hunk 10]
```diff
@@ -30,10 +29,12 @@ export default internalTask(
...
         if (!signer) [signer] = await ethers.getSigners();
-        instance = new Contract(contractaddress, abi, signer);
+        // instance = new Contract(contractaddress, abi, signer);
+        instance = await ethers.getContractAt(abi, contractaddress, signer);
       }
```
Comment (hilmarx) [inline buidler/tasks/internal/contracts/iTask.instantiateContract.js:38]:
```suggestion
```

File: buidler/tasks/gelato/core/task_receipts/iTask.fetchTaskReceipt.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (gitpusha) [reply buidler/tasks/gelato/core/task_receipts/iTask.fetchTaskReceipt.js:52]: On which network?  You would need to make sure that that network has the property: 
```
filters: { defaultFromBlock: 0 }
```
in its config

File: contracts/gelato_provider_modules/GelatoProviderModuleStandard.sol [hunk 11]
```diff
@@ -19,6 +19,16 @@ abstract contract GelatoProviderModuleStandard is IGelatoProviderModule {
...
+    function execPayload(uint256, address, address, Task calldata, uint256)
+        external
+        view
+        virtual
+        override
+        returns(bytes memory payload, bool)
+    {
+        return (payload, false);
+    }
+
```
[the remark this answers is not in the document]
Comment (gitpusha) [reply contracts/gelato_provider_modules/GelatoProviderModuleStandard.sol:31]: But this is already enforced. The `GelatoProviderModuleStandard` is an `abstract` contract which inherits from `IGelatoProviderModule` which has the `execPayload` fn on its interface. So developers always have to implement the `execPayload` method. But there is no standard implementation for it, which is why it's not part of the Standard.

File: deploy/setupGelatoCore.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  const executorSigner = ethers.provider.getSigner(executor);
+  const { deploy, log } = deployments;
+
+  // 1. Deploy Gelato Gas Price Oracle with gasPriceOracle => Chainlink on Mainnet
+  // @dev Acctoung "gasPriceOracle" can change gas prices for testing purposes
+  const GelatoGasPriceOracle = await deploy("GelatoGasPriceOracle", {
+    from: gasPriceOracle,
+    gas: 4000000,
+    args: [STARTING_GAS_PRICE],
+  });
```
[the remark this answers is not in the document]
Comment (gitpusha) [reply deploy/setupGelatoCore.js:37]: I would add an if clause then. Something like:

```
if (network.name === "buidlerevm" || network.name === "localhost")`
      deployGelatoGasPriceOracle
```

**Comment by gitpusha** [comment]
Closing because not going to use `buidler-deploy` in the foreseeable future (as discussed)

**Pull request closed without merging**